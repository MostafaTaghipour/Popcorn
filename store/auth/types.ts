import { Action } from "redux";
import { ApiAction, RequestState, PostRequestState } from "@app/types/action";

export interface AuthRequestBody {
  username: string;
  password: string;
}

export interface AuthResponseBody {
  token: string;
}

// auth state model
export interface AuthState {
  /**
   * authenticate request state
   *
   * @type {PostRequestState}
   * @memberof AuthState
   */
  authentication: PostRequestState;
  /**
   * user token
   *
   * @type {string}
   * @memberof AuthState
   */
  token?: string;
}

// actions
export enum AuthActionTypes {
  AUTHENTICATION = "AUTHENTICATION",
  SIGN_OUT = "SIGN_OUT",
}

export interface AuthenticationAction
  extends ApiAction<AuthActionTypes.AUTHENTICATION, AuthResponseBody> {}

export interface SignOutAction extends Action<AuthActionTypes.SIGN_OUT> {}

export type AuthActions = AuthenticationAction | SignOutAction;
