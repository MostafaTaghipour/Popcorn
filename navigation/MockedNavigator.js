import React from "react";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import { Provider as ReduxProvider } from "react-redux";
import { AuthProvider } from "../contexts/AuthContext";

const Stack = createStackNavigator();

/**
 * Provide a MockedNavigator to wrap our screens
 * 
 * Access navigation in screens 
 * 
 * @param {*} param0 
 */
export const MockedNavigator = ({ component, params = {}, store }) => {
  return (
    <ReduxProvider store={store}>
      <AuthProvider>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen
              name="MockedScreen"
              component={component}
              initialParams={params}
            />
          </Stack.Navigator>
        </NavigationContainer>
      </AuthProvider>
    </ReduxProvider>
  );
};
