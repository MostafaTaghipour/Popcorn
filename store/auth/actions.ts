import {
  AuthActionTypes,
  AuthenticationAction,
  AuthRequestBody,
  SignOutAction,
} from "./types";
import { AsyncActionStatus } from "@app/types/action";
import HttpStatusCode from "@app/types/net";
import { showToastAction } from "../ui/actions";
import LocaleManager from "@app/helpers/LocaleManager";
import AuthApi from "@app/data/remote/AuthApi";
import { AxiosError } from "axios";
import mocker from "@app/net/mocker";

/**
 * authenticate user action
 *
 * @param {string} username
 * @param {string} password
 * @return {*}
 */
export const authAsyncAction = (username: string, password: string) => {
  return async (dispatch: any) => {
    var action: AuthenticationAction = {
      type: AuthActionTypes.AUTHENTICATION,
      status: AsyncActionStatus.REQUEST,
    };

    dispatch(action);

    try {
      const body: AuthRequestBody = { username, password };
      const response = await AuthApi.authenticate(body);

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);

      if ((error as AxiosError)?.response?.status == HttpStatusCode.BadRequest)
        dispatch(
          showToastAction({
            text: LocaleManager.t("auth.wrong_credentials"),
            type: "danger",
          })
        );
    }
  };
};

export const signOutAction = (): SignOutAction => {
  return {
    type: AuthActionTypes.SIGN_OUT,
  };
};
