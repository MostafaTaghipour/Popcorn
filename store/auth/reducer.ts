import { AuthState, AuthActionTypes, AuthActions } from "./types";
import { Reducer } from "redux";
import createSecureStore from "redux-persist-expo-securestore";
import { persistReducer, PersistConfig } from "redux-persist";
import { AsyncActionStatus } from "@app/types/action";


const initialState: AuthState = {
  authentication: {
    error: undefined,
    inProgress: false,
    done: false,
  },

  token: undefined,
};

export const reducer: Reducer<AuthState, AuthActions> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case AuthActionTypes.AUTHENTICATION:
      switch (action.status) {
        case AsyncActionStatus.REQUEST:
          return {
            ...state,
            authentication: {
              ...state.authentication,
              inProgress: true,
              done: false,
            },
          };
        case AsyncActionStatus.SUCCESS:
          return {
            ...state,
            authentication: {
              ...state.authentication,
              inProgress: false,
              done: true,
              error: undefined,
            },
            token: action?.data?.token,
          };
        case AsyncActionStatus.FAILURE:
          return {
            ...state,
            authentication: {
              ...state.authentication,
              error: action.error,
              inProgress: false,
              done: false,
            },
          };
        default:
          return state;
      }

    case AuthActionTypes.SIGN_OUT:
      return initialState;

    default:
      return state;
  }
};

// we use secure storage to keep user creditionals safe
const secureStorage = createSecureStore();

// we persist user token
const persistConfig: PersistConfig<AuthState> = {
  key: "auth",
  storage: secureStorage,
  whitelist: ["token"],
};

export const authReducer = persistReducer(persistConfig, reducer);
