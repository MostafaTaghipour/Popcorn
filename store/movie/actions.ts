import {
  FetchCategoriesAction,
  MovieActionTypes,
  Movie,
  FetchCategoryMoviesAction,
  SearchMoviesAction,
  ClearSearchResultAction,
} from "./types";
import { AsyncActionStatus, PaginationRequestState } from "@app/types/action";
import MovieApi from "@app/data/remote/MovieApi";
import { store } from "..";
import Configs from "@app/constants/Configs";
/**
 * fetch movie categories action
 *
 * @return {*}
 */
export const fetchCategoriesAsyncAction = () => {
  return async (dispatch: any) => {
    var action: FetchCategoriesAction = {
      type: MovieActionTypes.FETCH_CATEGORIES,
      status: AsyncActionStatus.REQUEST,
    };

    dispatch(action);

    try {
      const response = await MovieApi.categories();

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data.results,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);
    }
  };
};

/**
 * fetch specific movie category movie action
 *
 * @return {*}
 */
export const fetchCategoryMovieAsyncAction = (
  category: string,
  pageNumber = Configs.PAGINATION_FIRST_PAGE_NUMBER,
  refreshing = false
) => {
  return async (dispatch: any) => {
    const currentState: PaginationRequestState<Movie[]> = store.getState().movie
      .currentCategoryMovies;

    //check if all data loaded
    if (currentState.data!.length >= currentState.total && !refreshing) return;

    //check if already loading
    if (currentState.loading || currentState.refreshing) return;

    //check is refreshing or loading more
    const page = refreshing ? Configs.PAGINATION_FIRST_PAGE_NUMBER : pageNumber;

    var action: FetchCategoryMoviesAction = {
      type: MovieActionTypes.FETCH_CATEGORY_MOVIES,
      status: AsyncActionStatus.REQUEST,
      payload: category,
      refreshing: refreshing,
      page: page,
    };
    dispatch(action);

    

    try {
      const limit = Configs.PAGINATION_PAGE_SIZE;
      const offset = page * limit;
      const response = await MovieApi.categoryMovies(category, limit, offset);

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data.results,
        total: response.data.count,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);
    }
  };
};

/**
 * search movie action
 *
 * @return {*}
 */
export const searchMovieAsyncAction = (
  query: string,
  pageNumber = Configs.PAGINATION_FIRST_PAGE_NUMBER,
  refreshing = false
) => {
  return async (dispatch: any) => {
    const currentState: PaginationRequestState<Movie[]> = store.getState().movie
      .searchResultMovies;

    //check if all data loaded
    if (currentState.data!.length >= currentState.total && !refreshing) return;

    //check if already loading
    if (currentState.loading || currentState.refreshing) return;

    //check is refreshing or loading more
    const page = refreshing ? Configs.PAGINATION_FIRST_PAGE_NUMBER : pageNumber;

    var action: SearchMoviesAction = {
      type: MovieActionTypes.SEARCH_MOVIES,
      status: AsyncActionStatus.REQUEST,
      refreshing: refreshing,
      page: page,
    };
    dispatch(action);

    try {
      const limit = Configs.PAGINATION_PAGE_SIZE;
      const offset = page * limit;
      const response = await MovieApi.search(query, limit, offset);

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data.results,
        total: response.data.count,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);
    }
  };
};

/**
 * clear old search results
 *
 * @return {*}  {ClearSearchResultAction}
 */
export const clearSearchResultsAction = (): ClearSearchResultAction => {
  return {
    type: MovieActionTypes.CLEAR_SEARCH_RESULTS,
  };
};
