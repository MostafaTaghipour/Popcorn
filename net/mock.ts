import { ListResult } from "./../store/movie/types";
/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */
import mocker from "./mocker";
import { mockCategories, mockMovies } from "@app/data/mock/movie";
import { mockAuthResponse } from "@app/data/mock/auth";

/**
 * we can define our mock data for each request
 *
 */
export const provideMockData = () => {
  mocker.addMock("user/auth-token", {
    data: (req: any) => {
      if ((req.data.username = "mostafa" && req.data.password == "123"))
        return mockAuthResponse;

      let err = Error("wrong credentials");

      //@ts-ignore
      err.code = 400;

      return err;
    },
    method: "post",
  });
  mocker.addMock("category", {
    data: { count: mockCategories.length, results: mockCategories },
    method: "get",
  });
  mocker.addMock("tags", {
    data: { count: mockMovies.length, results: mockMovies },
    method: "get",
  });
  mocker.addMock("search", {
    data: { count: mockMovies.length, results: mockMovies },
    method: "get",
  });
};
