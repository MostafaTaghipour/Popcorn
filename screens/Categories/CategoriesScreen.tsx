import { StyleSheet, View, SafeAreaView } from "react-native";
import React, { useEffect, useContext } from "react";
import {
  Title,
  ActivityIndicator,
  Divider,
  TouchableRipple,
} from "react-native-paper";
import { CategoriesStackNavProps } from "@app/navigation/categories.stack";
import { useDispatch } from "react-redux";
import { fetchCategoriesAsyncAction } from "@app/store/movie/actions";
import { FlatList } from "react-native-gesture-handler";
import { Category } from "@app/store/movie/types";
import { Feather } from "@expo/vector-icons";
import { useScrollToTop } from "@react-navigation/native";
import { useReduxSelector } from "@app/store";
import { ConnectivityContext } from "@app/contexts/ConnectivityContext";
import { showToastAction } from "@app/store/ui/actions";
import { LocalizationContext } from "@app/contexts/LocalizationContext";

/**
 * CategoriesScreen show all movie categories
 *
 * @export
 * @param {CategoriesStackNavProps<"Categories">} {
 *   navigation,
 * }
 * @return {*}
 */
export default function CategoriesScreen({
  navigation,
}: CategoriesStackNavProps<"Categories">) {
  const data = useReduxSelector((state) => state.movie.categories);
  const { isConnected } = useContext(ConnectivityContext);
  const { t } = useContext(LocalizationContext);
  const dispatch = useDispatch();
  const scrollRef = React.useRef(null);

  /**
   * scroll to top when tapping on the active tab
   * https://reactnavigation.org/docs/use-scroll-to-top/
   *
   */
  useScrollToTop(scrollRef);

  useEffect(() => {
    if (isConnected) dispatch(fetchCategoriesAsyncAction());
    else
      dispatch(
        showToastAction({
          text: t("error.no_internet"),
          type: "danger",
        })
      );
  }, []);

  /**
   * goto category detail and show category movies
   *
   * @param {Category} item
   */
  function categoryMovies(item: Category) {
    if (isConnected) navigation.navigate("Category", { title: item.name });
    else
      dispatch(
        showToastAction({
          text: t("error.no_internet"),
          type: "danger",
        })
      );
  }

  function renderItem(item: Category) {
    return (
      <CategoryItem category={item} onPress={(item) => categoryMovies(item)} />
    );
  }

  return (
    <SafeAreaView style={styles.container} testID="mySafe">
      {data.data && data.data.length > 0 ? (
        <FlatList
          style={styles.list}
          data={data.data}
          ref={scrollRef}
          testID="CategoryList"
          renderItem={({ item }) => renderItem(item)}
          ItemSeparatorComponent={() => <Divider />}
          keyExtractor={(item) => item.id.toString()}
        />
      ) : data.loading ? (
        <ActivityIndicator size="large" />
      ) : null}
    </SafeAreaView>
  );
}

/**
 *  CategoryItem pur component
 *
 *  @type {*} */
const CategoryItem = React.memo(
  ({
    category,
    onPress,
  }: {
    category: Category;
    onPress: (category: Category) => any;
  }) => {
    return (
      <TouchableRipple onPress={() => onPress(category)} testID="CategoryItem">
        <View style={styles.item}>
          <Title>{category.name}</Title>
          <Feather name="chevron-right" size={24} color="gray" />
        </View>
      </TouchableRipple>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
  list: {
    flex: 1,
    width: "100%",
  },
  item: {
    flexDirection: "row",
    justifyContent: "space-between",
    padding: 16,
  },
});
