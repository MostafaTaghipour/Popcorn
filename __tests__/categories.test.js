import React from "react";
import CategoriesScreen from "../screens/Categories/CategoriesScreen";
import { MockedNavigator } from "../navigation/MockedNavigator";
import { render } from "react-native-testing-library";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import mocker from "../net/mocker";
import { fetchCategoriesAsyncAction } from "../store/movie/actions";
import { MovieActionTypes } from "../store/movie/types";
import { reducer } from "../store/movie/reducer";
import { mockCategories } from "../data/mock/movie";
import { AsyncActionStatus } from "../types/action";

/**
 * test CategoriesScreen rendering
 */
describe("Categories render", () => {
  let initialState;
  let store;

  /**
   * initialize mock redux store
   */
  beforeAll(() => {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);

    // Initialize mockstore
    initialState = {
      movie: {
        categories: {
          data: mockCategories,
          loaded: true,
          loading: false,
        },
      },
      auth: {
        token: undefined,
      },
    };
    store = mockStore(initialState);
  });

  /**
   * Test CategoriesScreen snapshot
   */
  it("Categories snapshot", () => {
    const snap = render(
      <MockedNavigator component={CategoriesScreen} store={store} />
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });

  it("Testing FlatList", () => {
    const tree = render(
      <MockedNavigator component={CategoriesScreen} store={store} />
    );

    expect(tree.getAllByTestId("CategoryList").length).toBe(1);
    expect(tree.getAllByTestId("CategoryItem").length).toBe(
      mockCategories.length
    );
  });
});

describe("Categories redux state", () => {
  let initialState;
  let store;

  beforeEach(() => {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);

    // Initialize mockstore
    initialState = {
      movie: {
        categories: {},
      },
      auth: {
        token: undefined,
      },
    };
    store = mockStore(initialState);
  });

  beforeAll(() => {
    return mocker.enableMocking();
  });

  afterAll(() => {
    return mocker.disableMocking();
  });

  it("check FETCH_CATEGORIES action", async () => {
    const expectedActions = [
      {
        type: MovieActionTypes.FETCH_CATEGORIES,
        status: AsyncActionStatus.REQUEST,
      },
      {
        type: MovieActionTypes.FETCH_CATEGORIES,
        status: AsyncActionStatus.SUCCESS,
        data: mockCategories,
      },
    ];

    return store.dispatch(fetchCategoriesAsyncAction()).then(() => {
      // return of  actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it("check reducer on success", () => {
    expect(
      reducer(initialState.movie, {
        type: MovieActionTypes.FETCH_CATEGORIES,
        status: AsyncActionStatus.SUCCESS,
        data: mockCategories,
      })
    ).toEqual({
      categories: {
        data: mockCategories,
        loaded: true,
        loading: false,
      },
    });
  });

  it("check reducer on failed", () => {
    expect(
      reducer(initialState.movie, {
        type: MovieActionTypes.FETCH_CATEGORIES,
        status: AsyncActionStatus.FAILURE,
        error: Error("test"),
      })
    ).toEqual({
      categories: {
        loading: false,
        error: Error("test"),
      },
    });
  });
});
