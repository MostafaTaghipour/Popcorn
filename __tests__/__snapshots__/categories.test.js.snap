// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories render Categories snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    onInsetsChange={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
                "marginTop": 64,
              }
            }
          >
            <View
              collapsable={false}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <RCTSafeAreaView
                      emulateUnlessSupported={true}
                      style={
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        }
                      }
                      testID="mySafe"
                    >
                      <RCTScrollView
                        ItemSeparatorComponent={[Function]}
                        collapsable={false}
                        data={
                          Array [
                            Object {
                              "id": 0,
                              "name": "Drama",
                            },
                            Object {
                              "id": 1,
                              "name": "Action",
                            },
                            Object {
                              "id": 2,
                              "name": "Comedy",
                            },
                          ]
                        }
                        disableVirtualization={false}
                        getItem={[Function]}
                        getItemCount={[Function]}
                        horizontal={false}
                        initialNumToRender={10}
                        keyExtractor={[Function]}
                        maxToRenderPerBatch={10}
                        numColumns={1}
                        onContentSizeChange={[Function]}
                        onEndReachedThreshold={2}
                        onGestureHandlerEvent={[Function]}
                        onGestureHandlerStateChange={[Function]}
                        onLayout={[Function]}
                        onMomentumScrollEnd={[Function]}
                        onScroll={[Function]}
                        onScrollBeginDrag={[Function]}
                        onScrollEndDrag={[Function]}
                        removeClippedSubviews={false}
                        renderItem={[Function]}
                        renderScrollComponent={[Function]}
                        scrollEventThrottle={50}
                        stickyHeaderIndices={Array []}
                        style={
                          Object {
                            "flex": 1,
                            "width": "100%",
                          }
                        }
                        testID="CategoryList"
                        updateCellsBatchingPeriod={50}
                        viewabilityConfigCallbackPairs={Array []}
                        windowSize={21}
                      >
                        <View>
                          <View
                            onLayout={[Function]}
                            style={null}
                          >
                            <View
                              accessible={true}
                              focusable={true}
                              isTVSelectable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  false,
                                  undefined,
                                ]
                              }
                              testID="CategoryItem"
                            >
                              <View
                                style={
                                  Object {
                                    "flexDirection": "row",
                                    "justifyContent": "space-between",
                                    "padding": 16,
                                  }
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#000000",
                                        "fontFamily": "System",
                                        "fontWeight": "400",
                                        "textAlign": "left",
                                      },
                                      Array [
                                        Object {
                                          "color": "rgba(0, 0, 0, 0.87)",
                                          "fontFamily": "System",
                                          "fontWeight": "500",
                                          "textAlign": "left",
                                          "writingDirection": "ltr",
                                        },
                                        Array [
                                          Object {
                                            "fontSize": 20,
                                            "letterSpacing": 0.15,
                                            "lineHeight": 30,
                                            "marginVertical": 2,
                                          },
                                          undefined,
                                        ],
                                        Array [
                                          Object {
                                            "fontSize": 20,
                                            "letterSpacing": 0.15,
                                            "lineHeight": 30,
                                            "marginVertical": 2,
                                          },
                                          undefined,
                                        ],
                                      ],
                                    ]
                                  }
                                >
                                  Drama
                                </Text>
                                <Text />
                              </View>
                            </View>
                            <View
                              style={
                                Array [
                                  Object {
                                    "backgroundColor": "rgba(0, 0, 0, 0.12)",
                                    "height": 0.5,
                                  },
                                  undefined,
                                  undefined,
                                ]
                              }
                            />
                          </View>
                          <View
                            onLayout={[Function]}
                            style={null}
                          >
                            <View
                              accessible={true}
                              focusable={true}
                              isTVSelectable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  false,
                                  undefined,
                                ]
                              }
                              testID="CategoryItem"
                            >
                              <View
                                style={
                                  Object {
                                    "flexDirection": "row",
                                    "justifyContent": "space-between",
                                    "padding": 16,
                                  }
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#000000",
                                        "fontFamily": "System",
                                        "fontWeight": "400",
                                        "textAlign": "left",
                                      },
                                      Array [
                                        Object {
                                          "color": "rgba(0, 0, 0, 0.87)",
                                          "fontFamily": "System",
                                          "fontWeight": "500",
                                          "textAlign": "left",
                                          "writingDirection": "ltr",
                                        },
                                        Array [
                                          Object {
                                            "fontSize": 20,
                                            "letterSpacing": 0.15,
                                            "lineHeight": 30,
                                            "marginVertical": 2,
                                          },
                                          undefined,
                                        ],
                                        Array [
                                          Object {
                                            "fontSize": 20,
                                            "letterSpacing": 0.15,
                                            "lineHeight": 30,
                                            "marginVertical": 2,
                                          },
                                          undefined,
                                        ],
                                      ],
                                    ]
                                  }
                                >
                                  Action
                                </Text>
                                <Text />
                              </View>
                            </View>
                            <View
                              style={
                                Array [
                                  Object {
                                    "backgroundColor": "rgba(0, 0, 0, 0.12)",
                                    "height": 0.5,
                                  },
                                  undefined,
                                  undefined,
                                ]
                              }
                            />
                          </View>
                          <View
                            onLayout={[Function]}
                            style={null}
                          >
                            <View
                              accessible={true}
                              focusable={true}
                              isTVSelectable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  false,
                                  undefined,
                                ]
                              }
                              testID="CategoryItem"
                            >
                              <View
                                style={
                                  Object {
                                    "flexDirection": "row",
                                    "justifyContent": "space-between",
                                    "padding": 16,
                                  }
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#000000",
                                        "fontFamily": "System",
                                        "fontWeight": "400",
                                        "textAlign": "left",
                                      },
                                      Array [
                                        Object {
                                          "color": "rgba(0, 0, 0, 0.87)",
                                          "fontFamily": "System",
                                          "fontWeight": "500",
                                          "textAlign": "left",
                                          "writingDirection": "ltr",
                                        },
                                        Array [
                                          Object {
                                            "fontSize": 20,
                                            "letterSpacing": 0.15,
                                            "lineHeight": 30,
                                            "marginVertical": 2,
                                          },
                                          undefined,
                                        ],
                                        Array [
                                          Object {
                                            "fontSize": 20,
                                            "letterSpacing": 0.15,
                                            "lineHeight": 30,
                                            "marginVertical": 2,
                                          },
                                          undefined,
                                        ],
                                      ],
                                    ]
                                  }
                                >
                                  Comedy
                                </Text>
                                <Text />
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTScrollView>
                    </RCTSafeAreaView>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      pointerEvents="box-none"
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(224, 224, 224)",
                "flex": 1,
                "shadowColor": "rgb(224, 224, 224)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                MockedScreen
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
